src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -1023,7 +1150,10 @@ ZooKeeperPtr ZooKeeper::create(const Poco::Util::AbstractConfiguration & config,
 
 ZooKeeperPtr ZooKeeper::startNewSession() const
 {
-    auto res = std::shared_ptr<ZooKeeper>(new ZooKeeper(args, zk_log));
+    if (reconnect_task)
+        (*reconnect_task)->deactivate();
+
+    auto res = std::shared_ptr<ZooKeeper>(new ZooKeeper(args, zk_log, availability_zones, std::move(optimal_impl)));
```

Comment:
That works fine when we know all az at the start either from config or from active asking.

Imagine that we do not know az from config.
It works also fine when only one node of tree is unavailable. If that node from other az we connect to the same at the start.
If the node from the same az is unavailable, than our back ground task tracks that one unavailable node because AvailabilityZoneInfo::UNKNOWN < AvailabilityZoneInfo::Other. When that node is online we will connect to it.

But that wont work when we have two or all nodes unavailable at start, then  back ground task will tracks some one node. That node could be not from the local az. After back ground task switches us to that node, we are not continue to look further.

Reply:
> two or all nodes unavailable

Nothing works at all when 2 of 3 nodes are unavailable. Although your comment still makes sense for 5 nodes, it's not a usual usecase and we don't need it